### PR TITLE
Implement thread channel caching and accessor methods

### PR DESCRIPTION
**Pull request by skykatik**
**Description:**
Depends on Discord4J/discord-json#127
- Added methods for saving and accessing thread and thread's members. But I want to know whether or not to add `countThreads()`/`countMembersInThread()` methods
- Exposed methods to `EntityRetreiver` and `ThreadChannel`, `Guild`, updated  `RestChannel`, `RestGuild`
- Created new object `ThreadListPart` object which is used for the result in listing methods

**Justification:** Implement full support of threads for persisting

# API Changes
- Remove ChannelService#listActiveThreads: instead use GuildService#listActiveGuildThreads

### CODE REVIEW

File: common/src/main/java/discord4j/common/store/legacy/LegacyStoreLayout.java [hunk 1]
```diff
@@ -354,16 +365,14 @@ public Mono<Void> onChannelCreate(int shardIndex, ChannelCreate dispatch) {
             case GUILD_CATEGORY:
             case GUILD_NEWS:
             case GUILD_STORE:
-            case GUILD_STAGE_VOICE: return saveChannel(dispatch);
+            case GUILD_STAGE_VOICE: return saveChannel(dispatch.channel());
```
Comment (Azn9) [inline common/src/main/java/discord4j/common/store/legacy/LegacyStoreLayout.java:368]: For this method and the two below, you should add the two following fields to the inner enum "Type" (L. ~440) and add them in the switch
```Java
PUBLIC_THREAD(11),
PRIVATE_THREAD(12),
```

Comment (skykatik) [reply common/src/main/java/discord4j/common/store/legacy/LegacyStoreLayout.java:368]: I will add new types to enum, but Channels with these types can come in the `ChannelCreate`?

Comment (Azn9) [reply common/src/main/java/discord4j/common/store/legacy/LegacyStoreLayout.java:368]: MB, I misread where I commented, I thought we were in the "saveChannel" method, sorry 😅 
In that case, no, but on the onThreadCreateMethod, you'll need to change the switch

**Comment by HC-224** [comment]
Well I didn't think a PR like this existed and started it myself :sweat_smile: 

Is there anything holding this merge back to be available in snapshots?

**Review by quanticc — APPROVED** [review]
Thanks a lot!

Included some polishing commits and we'll monitor how it behaves in the 3.3.0 snapshots

**Comment by HC-224** [comment]
Thanks for merging this! I'll push this to my bot and bring up any issues if they arise.

**Comment by LatvianModder** [comment]
Found an issue: https://github.com/Discord4J/Discord4J/issues/1118

**Pull request merged**